pyro/infer/mcmc/hmc.py

```diff
@@ -251,14 +258,32 @@ def _initialize_model_properties(self, model_args, model_kwargs):
         self._prototype_trace = trace
 
     def _initialize_adapter(self):
-        mass_matrix_size = sum([p.numel() for p in self.initial_params.values()])
-        site_value = list(self.initial_params.values())[0]
-        if self._adapter.is_diag_mass:
-            initial_mass_matrix = torch.ones(mass_matrix_size,
-                                             dtype=site_value.dtype,
-                                             device=site_value.device)
+        if self._adapter.dense_mass is False:
+            dense_sites_list = []
+        elif self._adapter.dense_mass is True:
+            dense_sites_list = [tuple(sorted(self.initial_params))]
         else:
-            initial_mass_matrix = eye_like(site_value, mass_matrix_size)
+            msg = "full_mass should be a list of tuples of site names."
+            dense_sites_list = self._adapter.dense_mass
+            assert isinstance(dense_sites_list, list), msg
+            for dense_sites in dense_sites_list:
+                assert dense_sites and isinstance(dense_sites, tuple), msg
+                for name in dense_sites:
+                    assert isinstance(name, str) and name in self.initial_params, msg
+        diag_sites = tuple(sorted([name for name in self.initial_params
+                                   if name not in set().union(*dense_sites_list)]))
```

Comment:
nit: avoid computing `set().union()` each time through the loop, using `dense_sites_set`.